R/zchunk_LA114.wind-energy.R

```diff
@@ -28,43 +30,241 @@ module_energy_LA114.wind_DISABLED <- function(command, ...) {
+        L114.curr_region_RsrcCurves <- subset( L114.RsrcCurves_EJ_ctry_wind, GCAM_region_ID == region_ID )
+        L114.price <- min( L114.curr_region_RsrcCurves$base.price )
+        # Evaluate the supply curve until the supply doesn't seem to be changing more than a tolerance
+        # with a minimum number of evaluations to ensure we make progress into a very shallow curve as well.
+        L114.prev_supply <- 0
+        L114.min_calcs <- 100
+        repeat {
+          # Evaluate supply curves for all coutries in the region at the current price and sum up the supply
+          L114.SupplyPoints.currR <- data.frame( GCAM_region_ID=region_ID, price=L114.price, supply=0 )
+          for(i in 1:nrow(L114.curr_region_RsrcCurves) ) {
+            L114.SupplyPoints.currR$supply <- L114.SupplyPoints.currR$supply + do.call( evaluate_smooth_res_curve,
+                                                                                        c( L114.curr_region_RsrcCurves[ i, c( "curve.exponent", "mid.price", "base.price", "maxSubResource" ) ],
+                                                                                           L114.price ) )
+          }
+          regional_price_supply_points <- bind_rows( regional_price_supply_points, L114.SupplyPoints.currR )
+          if(L114.min_calcs <= 0 & ( L114.SupplyPoints.currR$supply - L114.prev_supply ) < L114.supply_tol ) {
+            # No longer making any difference in supply, we can stop
+            break
+          } else {
+            # Supply is still changing, need more price points
+            L114.min_calcs <- L114.min_calcs - 1
+            L114.price <- L114.price + L114.price_inc
+            L114.prev_supply <- L114.SupplyPoints.currR$supply
+          }
+        }
+      }
+
+    } else {
+
+    # The min and max price points for each region will be determined. Remember that the minimum prices are taken
+    # from the Zhou data. The maximum prices are calculated using the function, generate_max_prices.
+    L114.RsrcCurves_EJ_ctry_wind %>%
+      # Calculate max price using the function, generate_max_prices
+      mutate(max.price = generate_max_prices(base.price, mid.price, curve.exponent)) %>%
+      # Select minimum and maxium price for each GCAM region
+      group_by(GCAM_region_ID) %>% # Note that when only one group is selected, no need for "ungroup" afterwards
+      summarise(min.price = min(base.price), max.price = max(max.price)) ->
+      regional_min_max_price_points
+
+    # Supply curves for each GCAM region ID will be generated.
+         # First, a set of price points will be generated for each region and applied to each country.
+         # Supply will then be calculated at each price point for each country.
+         # Finally, supply will be summed for each GCAM region ID.
+    regional_price_supply_points <- lapply(regional_min_max_price_points$GCAM_region_ID, function(region_ID) {
```

Comment:
@pralitp and I tried using `group_by` and `mutate` first. If I remember correctly, we couldn't get the following batch of code to work without lapply or a loop. (Pralit, please correct me if I'm wrong.)

A unique sequence of price points is generated for each region, and then the existing table is expanded per the price points using `repeat_add_columns`. (This will then be used to calculate supply.) In any case, we might be able to cut off the `lapply` sooner. Or if you know of another way to write the following code:

```prices <- tibble(p = seq(data$min.price, data$max.price, L114.price_inc))

      # Applies the price points to each country (maintaining essential curve data)
      L114.RsrcCurves_EJ_ctry_wind %>%
        filter(GCAM_region_ID == region_ID) %>%
        select(curve.exponent, mid.price, base.price, maxSubResource) %>%
        repeat_add_columns(prices) ->
        ctry_curves```